lib/pbench/server/auth/auth.py

```diff
@@ -101,3 +106,56 @@ def verify_auth(auth_token):
                 e,
             )
         return None
+
+    @staticmethod
+    def verify_third_party_tokens(
+        auth_token: str, oidc_client: OpenIDClient, tokeninfo_endpoint: str = None
+    ) -> bool:
+        """
+        Verify the tokens provided to the Pbench server that were obtained by the
+        third party identity providers.
+        Args:
+            auth_token: Token to authenticate
+            oidc_client: OIDC client to call to authenticate the token
+            tokeninfo_endpoint: Optional tokeninfo_endpoint to validate
+                                tokens online in case offline verification
+                                results in some exception.
+        Returns:
+            JWT token string
+        """
+        # Verify auth token validity
+        identity_provider_pubkey = oidc_client.get_oidc_public_key(auth_token)
+        try:
+            oidc_client.token_introspect_offline(
+                token=auth_token,
+                key=identity_provider_pubkey,
+                audience=oidc_client.client_id,
+                options={
+                    "verify_signature": True,
+                    "verify_aud": True,
+                    "verify_exp": True,
+                },
+            )
+            return True
+        except (
+            jwt.ExpiredSignatureError,
+            jwt.InvalidTokenError,
+            jwt.InvalidAudienceError,
+        ):
+            Auth.logger.error("oidc token verification failed")
+        except Exception as e:
+            Auth.logger.exception(
+                "Unexpected exception occurred while verifying the auth token {!r}: {}",
+                auth_token,
+                e,
+            )
+            if not tokeninfo_endpoint:
+                Auth.logger.warning("Can not perform oidc online token verification")
+            else:
+                token_payload = oidc_client.token_introspect_online(
+                    token=auth_token, token_info_uri=tokeninfo_endpoint
+                )
+                if oidc_client.client_id in token_payload["aud"]:
+                    # If our client is an intended audience then only we accept the token
+                    return True
```

Comment:
If some failure occurs while introspecting the token online then the `_post` in `lib/pbench/server/auth/__init__.py`  will raise `OpenIDClientError` which uses `INTERNAL_SERVER_ERROR`. However, if that isn't handled flask will by default raise the same error anyway (like in this case) but I think It will be very rare. I am only introspecting online in case of the offline introspection results in some unexpected error. 

I do realize though that If the token is invalid then the online introspection will return a dictionary with just a message like `{msg: "Invalid Token"}` so I should check for the successes. Also for the failures, I think we don't need to let the exception blow by, I should catch them and let the user know what happened and why we could not verify the token.